### PR TITLE
[Ext/Decoder] Support person detection models for OpenVino @open sesame 8/18 13:26

### PR DESCRIPTION
**Pull request by wooksong**
This patch updates the tensor-decoder subplugin, "bounding-boxes", to support person detection models for OpenVino.

Signed-off-by: Wook Song <wook16.song@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed [ ]Skipped
2. Run test: [*]Passed [ ]Failed [ ]Skipped

#### How to test on AMD64 Ubuntu with mkldnn acceleration (CPU) ####

1. Build 
```console
$ pwd
./nnstreamer
$ sudo apt install openvino-dev
$ meson build -Denable-openvino=true
$ ninja -C build
```
2. Prepare the running environment
```console
$ export GST_PLUGIN_PATH=$(pwd)/build/gst/nnstreamer
$ export NNSTREAMER_CONF=$(pwd)/build/nnstreamer-test.ini
$ export NNSTREAMER_FILTERS=$(pwd)/build/ext/nnstreamer/tensor_filter
$ export NNSTREAMER_DECODERS=$(pwd)/build/ext/nnstreamer/tensor_decoder
```
3. Launch the pipeline

You can get the model files from [here](https://download.01.org/opencv/2019/open_model_zoo/R3/20190905_163000_models_bin/person-detection-retail-0013/FP32/).

```console
$ gst-launch-1.0 filesrc location=./src.png ! pngdec ! videoscale ! imagefreeze ! videoconvert ! video/x-raw,height=320,width=544,format=BGR,framerate=0/1 ! tensor_converter ! tensor_transform mode=typecast option=float32 ! tensor_transform mode=dimchg option=0:2 ! tensor_filter framework=openvino model=./person-detection-retail-0013.xml silent=false accelerator=true:cpu ! tensor_decoder mode=bounding_boxes option1=ov-person-detection option4=1200:675 option5=544:320 ! videoconvert ! video/x-raw,format=RGBA ! pngenc ! filesink location=./test.png
```
4. Results
- Source image
src

- The generated image containing bounding boxes
boxes

### CODE REVIEW

File: ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c [hunk 1]
```diff
@@ -115,15 +121,16 @@ typedef struct
...
-  {
-    properties_TFLite_SSD tflite_ssd; /**< Properties for tflite_ssd configured by option 1 + 3 */
+  struct {
+    union
+    {
+      properties_TFLite_SSD tflite_ssd; /**< Properties for tflite_ssd configured by option 1 + 3 */
+    };
+    /* From option2 */
+    imglabel_t labeldata;
+    char *label_path;
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c:131]: What's this internal struct??

Comment (wooksong) [reply ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c]: @myungjoo Corrected.

File: ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c [hunk 2]
```diff
@@ -919,13 +1016,34 @@ bb_decode (void **pdata, const GstTensorsConfig * config,
...
+        g_assert (0);
+    }
   } else {
     GST_ERROR ("Failed to get output buffer, unknown mode %d.", bdata->mode);
     return GST_FLOW_ERROR;
   }
 
   draw (&out_info, bdata, results);
-  g_array_free (results, TRUE);
+  g_array_free (results, FALSE);
```
Comment (myungjoo) [inline ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c:1046]: Why?

Comment (wooksong) [reply ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c:1046]: @myungjoo It seems that ```results``` holds its own memory spaces for each element and ```g_array_append_val``` just copies the data to the internal memory spaces. I think that it is enough to just call ```g_array_free``` without setting the second parameter unless we append a container of something.

File: ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c [hunk 3]
```diff
@@ -144,6 +153,8 @@ _init_modes (bounding_boxes * bdata)
     return TRUE;
   } else if (bdata->mode == TF_SSD_BOUNDING_BOX) {
     return TRUE;
+  } else if (bdata->mode == OV_PERSON_DETECTION_BOUNDING_BOX) {
+    return TRUE;
   }
   return TRUE;
```
Comment (gichan-jang) [inline ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c:159]: Always return TRUE?

File: ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c [hunk 4]
```diff
@@ -180,6 +192,7 @@ _exit_modes (bounding_boxes * bdata)
   if (bdata->mode == TFLITE_SSD_BOUNDING_BOX) {
     /* properties_TFLite_SSD *data = &bdata->tflite_ssd; */
   } else if (bdata->mode == TF_SSD_BOUNDING_BOX) {
+  } else if (bdata->mode == OV_PERSON_DETECTION_BOUNDING_BOX) {
```
Comment (gichan-jang) [inline ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c:195]: Why do you need this line?

**Review by gichan-jang — APPROVED** [review]
It works well in Ubuntu and RPI.
It would be good to choose the color of the bounding box according to the mode.

File: ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c (hunk 3 above)
Comment (wooksong) [reply ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c:159]: @gichan-jang As you pointed out, the lines at 156 ~ 157 and at 195 ~ 196 are not necessary. I will remove them.

File: ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c (hunk 4 above)
Comment (wooksong) [reply ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c:195]: @gichan-jang Please refer https://github.com/nnstreamer/nnstreamer/pull/2652#discussion_r472665486

File: ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c (hunk 2 above)
Comment (myungjoo) [reply ext/nnstreamer/tensor_decoder/tensordec-boundingbox.c:1046]: Ah.. got it..

**Review by myungjoo — APPROVED** [review]
LGTM

**Pull request merged**